File: 2020_November_Leetcode_30_days_challenge/Week_2_Poor Pigs/by_iteration.py
'''

Description:

There are 1000 buckets, one and only one of them is poisonous, while the rest are filled with water. They all look identical. If a pig drinks the poison it will die within 15 minutes. What is the minimum amount of pigs you need to figure out which bucket is poisonous within one hour?

Answer this question, and write an algorithm for the general case.

 

General case:

If there are n buckets and a pig drinking poison will die within m minutes, how many pigs (x) you need to figure out the poisonous bucket within p minutes? There is exactly one bucket with poison.

 

Note:

A pig can be allowed to drink simultaneously on as many buckets as one would like, and the feeding takes no time.
After a pig has instantly finished drinking buckets, there has to be a cool down time of m minutes. During this time, only observation is allowed and no feedings at all.
Any given bucket can be sampled an infinite number of times (by an unlimited number of pigs).

'''



class Solution:
    def poorPigs(self, buckets: int, minutesToDie: int, minutesToTest: int) -> int:
        
        pigs = 0
        while (minutesToTest / minutesToDie + 1) ** pigs < buckets:
            pigs += 1
        
        return pigs


# n : the number of buckets

## Time Complexity: O( n )
#
# The overhead in time is the cost of while-loop iteration, which is of O( n )

## Space Complexity: O( 1 )
#
# The overhead in space is the storage for loop index and temporary variable, which is of O( 1 )


import unittest

class Testing( unittest.TestCase ):

    def test_case_1( self ):

        result = Solution().poorPigs(buckets=1000, minutesToDie=15, minutesToTest=60 )
        self.assertEqual(result, 5)


if __name__ == '__main__':

    unittest.main()        

File: No_0509_Fibonacci Number/go_iteration/by_iteration.go
package No_0509

func fib(n int) int {

    var getFib func(int) int
    
    memo := map[int]int{0:0, 1:1}
    
    getFib = func(n int) int{
        
        // base cases
        if value, exist := memo[n]; exist == true{
            return value
        }
        
        // general cases
        fibN := getFib(n-1) + getFib(n-2)
        memo[n] = fibN
        
        return fibN
    }
    
    return getFib(n)
}



// n : the value of input

//// Time Complexity: O( n )
//
// The overhead in time is the cost of iteration, which is of O( n )

//// Sapce Complexity: O( 1 )
//
// The overhead in space is the storage for loop index and temporary varaible, which is of O( 1 )

// run "go test -v" in consonle to run unittest